### PR TITLE
[SYCL] Fix vec class alignment on windows platform

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
- For Windows alignment request for vector type larger than 64 is applied, but limited to 64

- Before this change, alignment request was ignored due to 64 byte from MSVC limit and default alignment was applied

- This default alignment could cause segmentation fault for 16-byte aligned memory access such as memory access for XMM

### CODE REVIEW

File: sycl/include/CL/sycl/types.hpp [hunk 1]
```diff
@@ -542,12 +542,13 @@ using vec_data_t = typename detail::vec_helper<T>::RetType;
...
 // Calling Convention
 // (https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention).
-#pragma message ("Alignment of class vec is not in accordance with SYCL \
+#pragma message("Alignment of class vec is not in accordance with SYCL \
 specification requirements, a limitation of the MSVC compiler(Error C2719).\
-Applied default alignment.")
-#define __SYCL_ALIGNAS(x)
+Requested alignment applied, limited at 64.")
+#define __SYCL_ALIGNED_VAR(type, x, var)                                       \
+  type __declspec(align((x < 64) ? x : 64)) var
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/types.hpp:549]: Can't we continue using  `alignas` specifier?

File: sycl/include/CL/sycl/types.hpp [hunk 2]
```diff
@@ -542,12 +542,13 @@ using vec_data_t = typename detail::vec_helper<T>::RetType;
 // For information on calling conventions for x64 processors, see
 // Calling Convention
 // (https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention).
-#pragma message ("Alignment of class vec is not in accordance with SYCL \
+#pragma message("Alignment of class vec is not in accordance with SYCL \
 specification requirements, a limitation of the MSVC compiler(Error C2719).\
-Applied default alignment.")
-#define __SYCL_ALIGNAS(x)
+Requested alignment applied, limited at 64.")
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/types.hpp:547]: @alexbatashev 
If I remember correctly you found that such a change is an ABI breaking one, could you please confirm?

Comment (alexbatashev) [reply sycl/include/CL/sycl/types.hpp:547]: It is

File: sycl/include/CL/sycl/types.hpp (hunk 1 above)
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/types.hpp:549]: `alignas` failed as follows  during kernel compilation for 'test_handler' from CTS.

```
In file included from C:\work\dongkyun\github\opensource\llvm_build\include\sycl\CL/sycl.hpp:11:
In file included from C:\work\dongkyun\github\opensource\llvm_build\include\sycl\CL/sycl/accessor.hpp:13:
In file included from C:\work\dongkyun\github\opensource\llvm_build\include\sycl\CL/sycl/buffer.hpp:11:
In file included from C:\work\dongkyun\github\opensource\llvm_build\include\sycl\CL/sycl/detail/buffer_impl.hpp:20:
C:\work\dongkyun\github\opensource\llvm_build\include\sycl\CL/sycl/types.hpp(1370,3): error: requested alignment is less than minimum alignment of 128 for type 'sycl::vec<double, 16>::DataType' (aka 'double __attribute__((ext_vector_type(16)))')
  __SYCL_ALIGNAS((detail::vector_alignment<DataT, NumElements>::value))
  ^
C:\work\dongkyun\github\opensource\llvm_build\include\sycl\CL/sycl/types.hpp(548,27): note: expanded from macro '__SYCL_ALIGNAS'
#define __SYCL_ALIGNAS(x) alignas((x<64)?x:64)
                          ^
```

File: sycl/include/CL/sycl/types.hpp (hunk 2 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/types.hpp:547]: Could you please remind why?

Comment (alexbatashev) [reply sycl/include/CL/sycl/types.hpp:547]: Type alignment is part of ABI: assembled code has no notion of structure layout, it operates raw bytes.

Comment (romanovvlad) [reply sycl/include/CL/sycl/types.hpp:547]: In this particular change I do not see how setting alignment can break anything:
1. The vector type is not passed directly to functions, but as part of the struct, so it's either passed by pointer or thru a register.
2. The structure has only one field which is already considered to be aligned by the clang as it's a vector type(generates vector instructions)
3. We set the alignment of vector type being equal to the size of the field itself, so we should not change the size. of the structure to trigger it to be passed differently.

**Review by romanovvlad — APPROVED** [review]
The patch doesn't fix the problem completely, compiling with `-march=skylake` triggers the issue even with the patch, but workaround the problem when compiling by "default".

**Pull request merged**